### ISSUE TITLE
KDA Chunk-wise Forward 算法探究报告

### ISSUE THREAD

**Issue by lingebeng** [lingebeng/op_dev#4]
# KDA Chunk-wise Forward 算法探究报告

## 一、KDA 基础递推公式（Recurrent 形式）

从 `naive_recurrent_kda` 可以提取出 KDA (Key-Decay Attention) 的逐步递推公式：

```
输入: q[B,T,H,K], k[B,T,H,K], v[B,T,H,V], g[B,T,H,K](log空间的遗忘门), beta[B,T,H]
状态: S[B,H,K,V] (隐藏状态矩阵)

对每个时间步 t:
  1. 衰减:     S = S * exp(g_t)[..., None]          # 按 K 维度的向量衰减
  2. Delta:    δ_t = v_t - (k_t^T @ S)               # Delta Rule 的 "纠错项"
  3. 更新:     S = S + beta_t * k_t ⊗ δ_t            # 外积更新
  4. 输出:     o_t = q_t^T @ S                        # 查询状态
```

**关键区别于 GLA**：
- GLA 的 gate `g` 是 **标量** (per-head)，KDA 的 gate `g` 是 **向量** (per-head-per-key-dim `[H, K]`)
- KDA 使用 **Delta Rule** 更新 (`v - k^T @ S`)，而非简单的 `k ⊗ v`
- KDA 额外引入 `beta` 作为更新的门控系数

---

## 二、Chunk-wise 分解算法（`naive_chunk_kda` 详解）

将长度 T 的序列切分为 `NT = T / BT` 个大小为 `BT` (默认64) 的 chunk。

### Phase 0: 预处理 — Gate Cumsum

```python
g = g.cumsum(-2)  # chunk 内 cumsum, shape [B,H,NT,BT,K]
```

这使得 chunk 内任意位置的相对衰减可以通过 `exp(g[...,i,:] - g[...,j,:])` 高效表示。

### Phase 1: 计算 Akk 矩阵 + 求逆（WY representation）

**目的**：将 Delta Rule 中隐式依赖的前缀序列转为显式矩阵形式。

```python
# Akk[..., t, j] = sum_d( k_t * exp(g_t - g_j) * k_j ) * beta_t
# 即 chunk 内 k 与 k 之间的 gated 内积矩阵 (下三角, 对角线 mask 掉)
A = compute_Akk(k, g, beta)    # [B,H,NT,BT,BT], 严格下三角

# 求解 (I - Akk)^{-1} 通过前向替代法
A_neg = -A.masked_fill(upper_mask, 0)
for i in range(1, BT):
    A[...,i,:i] += (A[...,i,:,None] * A[...,:,:i]).sum(-2)
A = (A + I) * beta[...,None,:]   # 最终得到 Akk_inv
```

**数学意义**：Delta Rule 中 `v_new = v - k @ S` 引入了隐式的 token 间依赖。`Akk_inv` (即代码中的 `A`) 本质上是 `(I - L)^{-1}` 的计算，其中 `L` 是 delta 更新矩阵的严格下三角部分。

### Phase 2: WY Representation — 计算 w 和 u

```python
w = Akk_inv @ (exp(g) * k)    # [B,H,NT,BT,K] — 用于状态的 "消除" 项
u = Akk_inv @ v                # [B,H,NT,BT,V] — 转换后的 value
```

- `u` 是 delta rule 解耦后的 "有效 value"
- `w` 是与隐藏状态 S 交互的 "有效 key"

### Phase 3: 计算 Aqk 矩阵（chunk 内 q-k 注意力）

```python
# Aqk[..., t, j] = sum_d( q_t * exp(g_t - g_j) * k_j )  (下三角, 含对角)
for j in range(BT):
    Aqk[..., j] = einsum('...cd, ...d -> ...c', q * exp(g - g_j), k_j)
Aqk = masked_fill(upper_strict_mask, 0)  # 因果掩码
```

### Phase 4: Inter-chunk 状态传播 + 输出（逐 chunk 串行）

```python
S = initial_state   # [B,H,K,V]
for i in range(NT):
    v_new_i = u_i - w_i @ S              # 减去状态对当前 chunk 的贡献
    o_i = (q_i * exp(g_i)) @ S + Aqk_i @ v_new_i   # inter + intra

    # 状态更新: 用 chunk 末尾的 gate 衰减，再加上 chunk 内的累积
    S = S * exp(g_last)                   # g_last = g_i[..., -1, :]
    S += (exp(g_last - g_i) * k_i)^T @ v_new_i
```

**输出分解**：
- **Inter-chunk 部分**: `(q_i * exp(g_i)) @ S` — 查询 chunk 之前积累的状态
- **Intra-chunk 部分**: `Aqk_i @ v_new_i` — chunk 内部的注意力

---

## 三、Triton 优化版的架构映射

`chunk_kda_fwd` (chunk_fwd.py) 将 naive 版本拆分成 **4 个独立 kernel 阶段**：

| 阶段 | 函数 | 对应 naive 的哪部分 |
|------|------|-------------------|
| ① Gate | `kda_gate_chunk_cumsum` 或 `chunk_local_cumsum` | `g.cumsum(-2)` + 可选的 gate 激活 |
| ② Intra | `chunk_kda_fwd_intra` | Aqk矩阵 + Akk矩阵 + 求逆 + WY (w, u) |
| ③ State | `chunk_gated_delta_rule_fwd_h` | 逐 chunk 的 S 传播 (v_new, h) |
| ④ Output | `chunk_gla_fwd_o_gk` | inter + intra 输出合成 |

### ② Intra 阶段进一步细分（sub-chunk BC=16）

大 chunk (BT=64) 被进一步分成 4 个 sub-chunk (BC=16)：

```
Step 2a: 对角块 (token-parallel)
  - chunk_kda_fwd_intra_token_parallel: 计算每个 16×16 对角块的 Aqk 和 Akk

Step 2b: 非对角块 + 合并求逆 (fused)
  - chunk_kda_fwd_kernel_inter_solve_fused:
    - 计算 16×16 sub-chunk 之间的交叉 Aqk 和 Akk 块
    - 在 sub-chunk 对角块上做前向替代
    - 用 block Neumann 级数合并为完整的 64×64 Akk_inv

Step 2c: WY
  - recompute_w_u_fwd: w = Akk_inv @ (exp(g) * k), u = Akk_inv @ v
```

### ③ State 阶段（来自 common/chunk_delta_h.py）

```
对每个 chunk i (串行):
  1. 保存 h[i] = S  (当前状态快照)
  2. v_new = u_i - w_i @ S    (减去状态贡献)
  3. S = S * exp2(g_last)     (gate 衰减)
  4. S += kg_i^T @ v_new      (chunk 内累积)
     其中 kg = k * exp(g_last - g)
```

### ④ Output 阶段（来自 gla/chunk.py 的 chunk_gla_fwd_kernel_o）

```
对每个 chunk i (可并行):
  o_i = scale * (q_i * exp2(g_i)) @ h[i]   # inter-chunk: 查询历史状态
      + Aqk_i @ v_new_i                     # intra-chunk: chunk 内注意力
```

---

### 推荐按以下函数逐步实现

```python
def chunk_kda_fwd_ref(q, k, v, g, beta, scale, initial_state, chunk_size=64):
    """
    JAX CPU reference for KDA chunk-wise forward.

    Args:
        q: [B, T, H, K]
        k: [B, T, H, K]
        v: [B, T, H, V]
        g: [B, T, H, K]  (log-space decay, 已经是最终 gate 值)
        beta: [B, T, H]
        scale: float
        initial_state: [B, H, K, V] or None
    Returns:
        o: [B, T, H, V]
        final_state: [B, H, K, V]
    """
```

**Step 1: Reshape + Gate Cumsum**

```python
# reshape: [B, T, H, K] -> [B, H, NT, BT, K]
# 在 chunk 维度做 cumsum
g_cumsum = jnp.cumsum(g_chunked, axis=-2)  # [B, H, NT, BT, K]
```

**Step 2: 计算 Akk 矩阵并求逆**

```python
def compute_Akk_and_invert(k, g, beta, BT):
    # Akk[..., t, j] = sum_d(k_t * exp(g_t - g_j) * k_j) * beta_t
    # 严格下三角
    # 前向替代求 (I - Akk)^{-1}
    # 最终 A = (A_inv + I) * beta
```

**Step 3: 计算 w, u (WY Representation)**

```python
w = A @ (jnp.exp(g) * k)    # [B, H, NT, BT, K]
u = A @ v                    # [B, H, NT, BT, V]
```

**Step 4: 计算 Aqk 矩阵**

```python
def compute_Aqk(q, k, g, BT):
    # Aqk[..., t, j] = sum_d(q_t * exp(g_t - g_j) * k_j) * scale
    # 下三角 (含对角)
```

**Step 5: Inter-chunk 传播 + 输出**

```python
# 用 jax.lax.scan 串行扫描 chunks
def scan_fn(S, chunk_data):
    q_i, k_i, u_i, g_i, w_i, Aqk_i = chunk_data
    v_new = u_i - w_i @ S
    o_i = (q_i * jnp.exp(g_i)) @ S * scale + Aqk_i @ v_new
    g_last = g_i[..., -1:, :]
    S = S * jnp.exp(g_last[..., 0, :, None])  # broadcast [K] -> [K,V]
    S += jnp.einsum('...ck,...cv->...kv',
                     jnp.exp(g_last - g_i) * k_i, v_new)
    return S, o_i

final_state, o = jax.lax.scan(scan_fn, S0, chunk_data)
```

### 关键注意事项

1. **exp2 vs exp**：Triton 优化版用 `exp2`（log2 基底）配合 `RCP_LN2` 缩放因子来提高精度/性能。CPU ref 直接用 `jnp.exp` 即可，但输入 gate 需要注意是否已经乘过 `1/ln2`。

2. **前向替代法的实现**：`(I - L)^{-1}` 的计算是 chunk 内的顺序操作，不可并行。在 JAX 中可以用 `jax.lax.fori_loop` 实现。

3. **Akk 的 sub-chunk 分块**：CPU ref 不需要 sub-chunk (BC=16) 的分块优化，直接在 BT×BT 粒度操作即可。

4. **gate 激活**：如果 `use_gate_in_kernel=True`，gate 计算为：
   ```python
   g = -exp(A_log) * softplus(g_raw + dt_bias)  # 标准模式
   # 或
   g = lower_bound * sigmoid(exp(A_log) * (g_raw + dt_bias))  # lower_bound 模式
   ```
   CPU ref 可以预先计算好 gate 值再传入。

5. **数值稳定性**：注意 `exp(g_i - g_j)` 中，由于 g 是 cumsum 后的值，差值天然有界。

6. **形状约定**：naive 版本使用 `[B, H, NT, BT, ...]` 布局，你的 JAX 版本可以保持一致或按习惯调整。

### 建议的测试验证路线

```
naive_recurrent_kda (逐步, ground truth)
    ↕ 对比
naive_chunk_kda (chunk naive ref, 已验证等价)
    ↕ 对比
你的 JAX chunk ref
```

---

## 五、与 GLA chunk 算法的关系

KDA 的 chunk-wise 算法与 GLA 共享了大量基础设施：

| 组件 | GLA | KDA | 差异点 |
|------|-----|-----|--------|
| Gate | 标量 `g[B,T,H]` | 向量 `g[B,T,H,K]` | KDA 每个 key dim 独立衰减 |
| 状态更新 | `S += k ⊗ v` | `S += β·k ⊗ (v - k^T@S)` | KDA 使用 Delta Rule |
| Akk 矩阵 | 不需要 | 需要 + 求逆 | Delta Rule 引入 token 间依赖 |
| Inter-chunk h 传播 | `chunk_gla_fwd_h` | `chunk_gated_delta_rule_fwd_h` | 共用同一个 kernel，输入不同 |
| Output kernel | `chunk_gla_fwd_o_gk` | 同 | 完全共用 |

**结论**：如果你已经实现了 GLA 的 chunk-wise，KDA 的核心增量是：
1. Gate 从标量变向量（影响 cumsum 和 exp 计算维度）
2. 增加 Akk 矩阵计算和求逆（WY representation）
3. `w, u` 的计算（替代直接使用 k, v）

---

## 六、源码文件索引

| 文件路径 | 内容说明 |
|---------|---------|
| `fla/ops/kda/naive.py` | naive_recurrent_kda + naive_chunk_kda (参考实现) |
| `fla/ops/kda/chunk.py` | ChunkKDAFunction (autograd wrapper) |
| `fla/ops/kda/chunk_fwd.py` | chunk_kda_fwd (forward 编排器) |
| `fla/ops/kda/chunk_intra.py` | Intra-chunk Triton kernels (Aqk, Akk, 求逆) |
| `fla/ops/kda/chunk_intra_token_parallel.py` | Token-parallel 对角块计算 |
| `fla/ops/kda/wy_fast.py` | WY representation (w, u 计算) |
| `fla/ops/kda/gate.py` | Gate 激活 + chunk cumsum |
| `fla/ops/kda/fused_recurrent.py` | Fused recurrent Triton kernel |
| `fla/ops/common/chunk_delta_h.py` | Inter-chunk 状态传播 (共用) |
| `fla/ops/gla/chunk.py` | Output kernel chunk_gla_fwd_o_gk (共用) |
